lib/discover/client_udp.go

```diff
@@ -113,12 +114,35 @@ func (d *UDPClient) broadcast() {
 	}
 
 	timer := time.NewTimer(0)
+
+	// Ugly hack with two goroutines to feed the trigger channel follows. This
+	// will be refactored when we move to the TLS based discovery thing, I
+	// promise - currently aiming for minimal changes here.
+
+	trigger := make(chan time.Time)
+
+	// Trigger fires when the timer does.
+	go func() {
+		for t := range timer.C {
+			trigger <- t
+		}
+	}()
+
+	// Trigger fires when the external port mapping changes.
+	eventSub := events.Default.Subscribe(events.ExternalPortMappingChanged)
+	defer events.Default.Unsubscribe(eventSub)
+	go func() {
+		for range eventSub.C() {
+			trigger <- time.Now()
+		}
+	}()
+
```

Comment:
I did that first, but sendAnnoncements would depend on a bunch of variables defined above the loop and needs parameters or them moved into the DiscoveryClient and I got tired of it halfway through fixing that. :)